autoload/ale/lsp_linter.vim

```diff
@@ -271,6 +271,29 @@ function! ale#lsp_linter#OnInit(linter, details, Callback) abort
         call ale#lsp#NotifyForChanges(l:conn_id, l:buffer)
     endif
 
+    " Tell the relevant buffer that the LSP has started via an autocmd.
+    if l:buffer > 0
+        if l:buffer == bufnr('')
+            silent doautocmd <nomodeline> User ALELSPStarted
+        else
+            execute 'augroup ALELSPStartedGroup' . l:buffer
+                autocmd!
+
+                execute printf(
+                \   'autocmd BufEnter <buffer=%d>'
+                \       . ' doautocmd <nomodeline> User ALELSPStarted',
```

Comment:
You want to run two commands on `BufEnter`:

1. `doautocmd`…
2. Remove the autocommand

Otherwise you haven't actually replicated `++once` correctly. The placement of the other autocmd means it's deleted before it even runs (though the syntax was wrong).

If we get rid of all the execute/printfs, the template we want is one of

```vim
augroup ALELSPStartedGroup
  autocmd!
  autocmd BufEnter <buffer=N> doautocmd <nomodeline> User ALELSPStarted | autocmd! ALELSPStartedGroup <buffer=N>
augroupEnd
```

_or_

```vim
augroup ALELSPStartedGroup
  autocmd!
  autocmd BufEnter <buffer=N> doautocmd <nomodeline> User ALELSPStarted
  autocmd BufEnter <buffer=N> autocmd! ALELSPStartedGroup <buffer=N>
augroup END
```

Either one is fine. Then just execute/printf to substitute `N` with the `l:buffer`.

Reply:
(See another comment of mine about the group name: it actually has the buffer number attached!)